### PR TITLE
client/asset/eth: finish client wallet tokens

### PR DESCRIPTION
**Pull request by buck54321**
I've peeled a few other PRs off of this one, but I think I'll submit the rest as one. Lots of tests written or adapted and all are passing. Ready for a look.

```
Introduce tokenContractor interface to handle operations on the token
contract and token-bound swap contract.

Separate ExchangeWallet into baseWallet and AssetWallet.

Move contractor handling to AssetWallet. This simplifies the ethFetcher
and is a better division of labor, making future implementations of
EVM-family assets easier.

Implement detailed calculations of swap and approve gas, issuing
warnings when a configuration mismatch is detected.

dex/networks/eth,dex:
Simplify dex.Token and create dexeth.Token type to add gas and address
info. Move gas info to version-specific token data.

```

### CODE REVIEW

File: client/asset/eth/eth.go
Comment (martonp) [inline client/asset/eth/eth.go:775]: I haven't read the whole PR yet, but I'm assuming that on the first sell trade of the token, this is where the approval would happen. If this is the case then there would need to be additional handling on the server, because this trade cannot yet be booked. Having the unlimited approval as a manual step completely avoids this issue. I know you think this is a bad user experience, but anyone who has used an erc20 token is completely familiar with this step. It's just an extra click before you sell the token.

Also, I think these checks of the allowance going below 2^256/2 are not necessary. Other than maybe some absolute ridiculous token, there will be no tokens anywhere near 2^256 max supply, and even for that ridiculous token, the current user would have had to initiate swaps with the entire supply of the token before their allowance would run out, so I don't think we need to worry about that.

**Review by martonp — COMMENTED** [review]
Initial Pass, looks good. I like the base/asset wallets and the contractors being moved out of the nodeclient. I think the biggest issue I found is regarding the unit conversion when initiating swaps.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:775]: > there would need to be additional handling on the server, because this trade cannot yet be booked

Can you expand on this? Doesn't the nonce order guarantee that the approval will be processed first.

From what I can tell, we're not checking the client's allowance server-side, and I didn't think we had any plans to do that. I was just going to relegate it to the client as part of configuration. Nothing we could do can prevent a client from sending an `approve(contractAddr, 0)` through after we accept their order anyway. 
The cost of approval is next to nothing, compartively, and correctly reported in the pre-order estimates.

If we did want to check allowance server-side, we would have to add latency handling though, which is what I think you were getting at.

To add manual approval, we would need to add a new UI dialog, new endpoint in webserver, new exported method from `*Core`, new exported method from `eth.ExchangeWallet`, and a new method or specialty interface in **interface.go**. And then we need to maintain all of that forever. And that's so we can force a user to click a button that they have no choice but to click in order to use the wallet that they just created. We also have to explain why we are forcing them to separately click this button to use the wallet. Why would they click "Create" if they were then going to say no to configuring the wallet for use? It makes for a bad UX, imo. Approval is not optional. It's part of trading with this asset. We don't need to pass the burden on to the user.

File: client/asset/eth/contractor.go
Comment (buck54321) [reply client/asset/eth/contractor.go:143]: Ah. Damn.

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:775]: Hm I'm probably  biased because all the other apps I've used have had approval as a manual step.  I think you're right, we don't need to do any validation on the server side. 

I think the ideal way would be the approval happens automatically when the user is doing their first buy order.  In that case this code that does the approval should be in `ReserveN` instead of `FundOrder`.

**Comment by chappjc** [comment]
I just noticed that dex/networks/eth/params_test.go is lacking an lgpl tag now.  Can add that too?

File: client/asset/eth/eth.go
Comment (chappjc) [inline client/asset/eth/eth.go:972]: The %d verbs don't have matching args.

File: client/asset/eth/eth.go
Comment (chappjc) [inline client/asset/eth/eth.go:964]: What do you mean by worry about accuracy?  As in just using the hard coded and padded estimates for this contract version?

**Review by chappjc — CHANGES_REQUESTED** [review]
Looking pretty slick so far.  Still going through it.

File: client/asset/eth/eth.go
Comment (JoeGruffins) [inline client/asset/eth/eth.go:665]: Still need to hold the lock to do `w.unlockFunds` I think.

File: client/asset/eth/eth.go
Comment (JoeGruffins) [inline client/asset/eth/eth.go:638]: Still need to hold the lock to do both locking and unlocking I think.

File: client/asset/eth/eth.go
Comment (JoeGruffins) [inline client/asset/eth/eth.go:1868]: Expensive.

**Review by JoeGruffins — COMMENTED** [review]
Looks good. I hope that you can add a `TokenWallet` similar to #1404 because I think that would clean up these methods a lot and make them easier to reason about without all the `w.assetID == BipID` checks.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:232]: There is only one `baseWallet` shared by all `AssetWallet`s. 

https://github.com/decred/dcrdex/blob/64b5c1e55e9a75937c4bd3d4f523bda17fff3813/client/asset/eth/eth.go#L512

I am going to look at `ETHWallet` and `TokenWallet` structs as per @JoeGruffins request though.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:665]: Whoops

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:232]: > There is only one `baseWallet` shared by all `AssetWallet`s.

Yah, this was really what I was commenting about.  Made me uncertain on why there where two separate types.  Will see what happens with the next iteration.

File: client/asset/eth/contractor.go
Comment (martonp) [inline client/asset/eth/contractor.go:444]: Use `evmify` instead of `WeiToGwei`

**Review by martonp — COMMENTED** [review]
Started reviewing, will continue later.

File: client/asset/eth/eth.go
Comment (chappjc) [inline client/asset/eth/eth.go:1117]: Why not use `w.parent` when accessing the mutex too, and eliminate the `ethWallet` definition?

File: client/asset/interface.go
Comment (chappjc) [inline client/asset/interface.go:490]: Only *one*

**Review by chappjc — APPROVED** [review]
I really like the latest change of splitting `ETHWallet` and `TokenWallet`.  In general it makes things much easier to follow and less error prone, even if at the expense of a little extra code and boiler plate.  Worth it.

File: client/asset/eth/eth.go
Comment (JoeGruffins) [inline client/asset/eth/eth.go:2638]: These three look like they should be off of `TokenWallet`.

File: client/asset/eth/eth.go
Comment (JoeGruffins) [inline client/asset/eth/eth.go:2703]: Comment is unexported, but this is exported. Will this be part of another tool? Does it belong in testing? Uses waitForMined() which is part of the harness tests. Not sure if you can use functions from test files in the build.

**Review by JoeGruffins — COMMENTED** [review]
Just a few more comments.

The eth harness tests are failing with `participant newV0TokenContractor error: error reading bound token address: no contract code at given address`

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:1390]: I think @buck54321 was posing that question.  I suppose the answer would just be convenience for core so it wouldn't have to figure out which ones to separate into different calls, if that's even something it should be aware of at all which I don't think it should.

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:2018]: I'm not sure I see much reason to allow locking an eth account at all.  As far as I'm concerned, they can be unlocked permanently when user logs in and the native eth wallet is loaded.  The app pass is still required to do various things, and always for withrdaw.

Alternatively, for a gentle solution to this `TODO` comment, how about overriding the `Lock` method with a dummy `func (*TokenWallet).Lock() error { return nil }`.

File: client/asset/eth/eth.go
Comment (martonp) [reply client/asset/eth/eth.go:241]: It's because there's only one instance of `baseWallet` that holds the reference to the node. Then there are multiple  instances of `AssetWallet` each of which hold a reference to the one `baseWallet`. It's not that each `AssetWallet` is holding a reference to it's own `baseWallet`.

**Review by martonp — APPROVED** [review]

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:1401]: There's already an unexported `redeem` method on `AssetWallet`.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:1662]: `AssetWallet` doesn't have access to the token parent.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:2410]: There was too much shared here. I didn't want to duplicate all of it.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:2703]: `waitForMined` is passed in as an argument for now. This will be used to create the gas tables once a contract is launched on a network. I haven't written a utility for it yet, but I'm planning on it.

File: client/asset/eth/eth.go
Comment (buck54321) [reply client/asset/eth/eth.go:2638]: Damn. I can't do this because of some other shared code. Maybe it would be worth splitting more methods up, but I'm not feeling it. Can follow up.

File: client/asset/eth/eth.go
Comment (chappjc) [reply client/asset/eth/eth.go:539]: I was being seriously dense about this too.

**Comment by chappjc** [comment]
Thanks for the hard work guys.  I think we can move forward.  @JoeGruffins If there's anything else that pops out or looks to have slipped by, there's clearly gonna be followup.

**Pull request merged**
